### ISSUE TITLE
Index IPNS

### ISSUE THREAD

**Issue by dokterbob** [ipfs-search/ipfs-search#48]
Index IPNS entries. I guess the sensible way would be to create an IPNS index with the IPNS id (pubkey) as ES document id. The document structure could be:

```json
{
  'links': {
    {
      'first-seen': '<date>',
      'last-seen': '<date>',
      'cid': '<ipfs-cid>'
    }
  }
}
```
This way we not only store the current state but also previous versions and when they were seen.

@madnificent Does this make sense?

**Comment by madnificent** [comment]
I believe it would also make sense to add the IPNS addresses to the indexed IPFS pages.  This would make it easy to figure out the IPNS link for an IPFS address, and show a list of variations.  Would that be plausible?

**Comment by dokterbob** [comment]
Yes, but towards the future I'm considering to separate data by how it's updated. This is because ES updates the whole document every time. I've already noticed the last seen has a fairly high impact on disk space. 

In the case of IPNS we could just as easily search through yet another index, plus we could do reverse lookups on results asynchronously, in separate API calls.

Does that sound reasonable? 
-- 
Verstuurd vanaf mijn Android apparaat met K-9 Mail. Excuseer mijn beknoptheid.

**Comment by madnificent** [comment]
It sounds reasonable either case, though I'm not sure I understand correctly.

My assumption now is that the IPNS links of an IPFS document would rarely be updated if we keep historical IPNS links. As in, an IPNS link may point to a new IPFS document often, but an IPNS document would then rarely receive a new IPNS link.

Under that assumption, it is smart to add the IPNS link to the IPFS ES document.  Any extra information regarding an IPNS link would better be stored in a separate ES index.

Is that in line with what you suggested?

**Comment by dokterbob** [comment]
No, I suggested just having a single place to store IPNS links, as a separate index. It would search just as easily, would expose full history, and would scale better as reverse lookups (what links here) would be asynchronous. 

Aad Versteden <notifications@github.com> schreef op 24 januari 2018 12:25:55 GMT+00:00:
>It sounds reasonable either case, though I'm not sure I understand
>correctly.
>
>My assumption now is that the IPNS links of an IPFS document would
>rarely be updated if we keep historical IPNS links. As in, an IPNS link
>may point to a new IPFS document often, but an IPNS document would then
>rarely receive a new IPNS link.
>
>Under that assumption, it is smart to add the IPNS link to the IPFS ES
>document.  Any extra information regarding an IPNS link would better be
>stored in a separate ES index.
>
>Is that in line with what you suggested?

-- 
Verstuurd vanaf mijn Android apparaat met K-9 Mail. Excuseer mijn beknoptheid.

**Comment by madnificent** [comment]
Ok, that works for me too.
